### ISSUE TITLE
Checking in package-lock.json?

### ISSUE THREAD

**Issue by jschfflr** [puppeteer/puppeteer#7239]
At the moment, we have the `package-lock.json` file in `.gitignore` and are instead relying on workarounds like https://github.com/puppeteer/puppeteer/pull/7238. We should documenten why we chose to do so or maybe reiterate that decision. 
@mathiasbynens: Are there historic reasons for this decision?

**Comment by mathiasbynens** [comment]
Echo'ing https://github.com/puppeteer/puppeteer/pull/7232#issuecomment-839834117, I'd like to avoid checking in `package-lock.json` if we can, because of the additional churn it introduces.

#7238 is _not_ a workaround for a missing `package-lock.json` — it's orthogonal, since every `npm install` [potentially](https://stackoverflow.com/questions/44297803/what-is-the-role-of-the-package-lock-json#comment78463335_44297998) updates `package-lock.json` (unless we pin versions in `package.json`, which is what #7238 checks).
In other words, even if we were to adopt `package-lock.json`, we'd still want to keep #7238.

If we were to adopt `package-lock.json` _without_ #7238, we’d have to switch to using `npm ci` instead of `npm install`. This is easy enough to do for our automated scripts, but it would be tough to ask every single Puppeteer contributor to change their workflow as well.

WDYT?

**Comment by jschfflr** [comment]
That sounds reasonable to me. WDYT about adding an explanation for this in the contributing guide for future reference?

**Comment by wojtekmaj** [comment]
I still think #7238 *is* a workaround, and in my opinion not a very good one. 

Pinning versions in `package.json` isn't an alternative to checking out `package-lock.json`. These dependencies have dependencies on their own, which are NOT pinned, and these dependencies have theirs, and so on. So in fact, with every install, you will end up with secondary, tertiary dependencies all different every time you install the project. Therefore, statement in #7238:

> "Because we don't check our package-lock.json in, we can end up with different versions installed locally vs CI (...) pin every version to avoid this.

is false. This doesn't fix anything and only gives a false impression of security/stability.

To avoid the churn of having `package-lock.json` unintentionally updated, I'd rather consider something like a GitHub Action that checks if `package-lock.json` was modified when `package.json` was not and commit message does not mention `update dependencies` or something like that.

Lastly, pinning version numbers in puppeteer causes lots of duplicated dependencies in my lockfile, e.g. I already had `progress@2.0.3` but since `puppeteer` requires `progress@2.0.1` and not `progress@^2.0.1` that version was not matching and it had to create a duplicate.

**Comment by foolip** [comment]
I've come across this issue because pinned dependencies create a problem for dependents of Puppeteer. In https://github.com/foolip/mdn-bcd-collector/pull/1537 I wanted to upgrade the node-fetch dependency 2.6.5 but found that Puppeteer's would not be deduplicated and get a copy of 2.6.1. This is the output of `npm ls node-fetch` I saw:
```
mdn-bcd-collector@3.3.1 /Users/foolip/mdn/mdn-bcd-collector
├─┬ @google-cloud/logging-winston@4.1.1
│ ├─┬ @google-cloud/logging@9.6.0
│ │ ├─┬ gcp-metadata@4.2.0
│ │ │ └─┬ gaxios@3.2.0
│ │ │   └── node-fetch@2.6.5  deduped
│ │ └─┬ google-gax@2.25.1
│ │   ├─┬ google-auth-library@7.9.1
│ │   │ └─┬ gaxios@4.3.1
│ │   │   └── node-fetch@2.6.5  deduped
│ │   └── node-fetch@2.6.5  deduped
│ └─┬ google-auth-library@7.0.4
│   └─┬ gaxios@4.2.1
│     └── node-fetch@2.6.5  deduped
├─┬ @google-cloud/storage@5.14.4
│ ├─┬ @google-cloud/common@3.7.2
│ │ └─┬ teeny-request@7.0.1
│ │   └── node-fetch@2.6.5  deduped
│ └─┬ gcs-resumable-upload@3.3.1
│   └─┬ gaxios@4.3.2
│     └── node-fetch@2.6.5  deduped
├─┬ @octokit/rest@18.11.2
│ └─┬ @octokit/core@3.5.1
│   └─┬ @octokit/request@5.6.1
│     └── node-fetch@2.6.5  deduped
├── node-fetch@2.6.5
└─┬ puppeteer@10.4.0
  └── node-fetch@2.6.1
```

Note how all of the dependencies except Puppeteer's are deduped.

Looking inside node_modules/puppeteer/node_modules/ I see the same thing has happened for the packages debug, progress and rimraf. The progress case is the simplest to illustrate with:
```
mdn-bcd-collector@3.3.1 /Users/foolip/mdn/mdn-bcd-collector
├─┬ eslint@7.32.0
│ └── progress@2.0.3
└─┬ puppeteer@10.4.0
  └── progress@2.0.1
```

I think there are two ways this could be improved:
- Aggressively update pinned dependencies, so that Puppeteer rarely has an older dependency than other packages.
- Revert the "dependencies" section of https://github.com/puppeteer/puppeteer/pull/7238

If you go back to ^ versions, perhaps a CI step could be added to run all the tests both with the earliest and the latest dependencies possible.

**Comment by foolip** [comment]
I've sent https://github.com/puppeteer/puppeteer/pull/7620 to configure Dependabot, if that's something this project would like. I use it on all my node projects and am pretty happy with it. The only downside is it creates a lot of PRs and makes `git log` noisy.

**Comment by wojtekmaj** [comment]
Dear maintainers, the recent faker/colors.js incident should have convinced you to check in your package-lock.json.

**Comment by wojtekmaj** [comment]
Bump
